### PR TITLE
feat(build): issues/28 docs build setup and repo restructure

### PR DESCRIPTION
**Pull request by cdcabrera**
## What's included
- feat(build): issues/28 docs build setup
  * initial repo build
- DEV GitHub PR and Issue templates
  * github pr and issue templates
- DEV build update
  * initial build tooling, testing
- DEV modules_docs adoc files
  * move adoc files into modules_docs
- DEV src directory layout
  * base src directory and symlinks
- DEV build tooling updates
  * tooling updates based on new src directory layout
  * NPM index.js file
  * base dist files set
  * local run development through docker setup
- DEV clean up symlinks, dist, locale script
- ~DEV move original RST files ~
  * ~relocated original rst files under src/rst~
- ~DEV update makefile RST~
- ~DEV relocate asciidoc-conversion files to src modules ~
  * ~relocate asciidoc-conversion files to src modules~
- ~DEV base build for asciidoctor and embedded assistance ~
  * ~build, linter, testing, packages, npm and build scripts~
  * ~build, initial jenkins and travis setup~
  * ~dist, initial output generated~
- ~DEV GitHub PR and Issue templates~

### Notes

- Went ahead and brought in both @jlprevatt and @annibond changes from their respective forks so we could get an idea of how to run the build. It is understood that by using that work, everything may be subject to alteration
- What should probably happen next is a discussion on what to change. We rearranged the basic structure to follow a `src` and `dist` structure, where `dist` contains the generated output.
  - Despite some reservations around checking them in (especially if on a Mac =)), it appears going the `symlink` method for linking Asciidoctor is the quickest route. It was easier to create soft symlinks on a Linux machine and port them over. The alternative was to end up copying the files into a temp structure, then running the same process, about double the effort. The basic directory structure we ended up with...
```
    - src
      - docs_install
        - discovery
          - dist [symlink]
          - modules [symlink]
          - shared [symlink]
          - docinfo.xml
          - master.adoc
        - qpc
          - dist [symlink]
          - modules [symlink]
          - shared [symlink]
          - docinfo.xml
          - master.adoc
      - docs_user
        - discovery
          - dist [symlink]
          - modules [symlink]
          - shared [symlink]
          - docinfo.xml
          - master.adoc
        - qpc
          - dist [symlink]
          - modules [symlink]
          - shared [symlink]
          - docinfo.xml
          - master.adoc
      - modules_docs
        - [asciidocs go here]
      - modules_ea
        - [JSON files go here]
      - shared
        - attributes_dsc.adoc
        - attributes_qpc.adoc
      - dist
        - install
          - index.html
          - index-brand.html
          - locales
            - en.json
            - locales.json [a list of all available locales in a GUI consumable format]
        - user
          - index.html
          - index-brand.html
        - index.js [an interface to allow GUI consumption]
      - scripts
        - [build scripts]
```

The previous dir structure...
```
    - src
      - docs_cli
        - discovery
          - dist [symlink]
          - modules [symlink to modules_docs]
          - attributes.adoc
          - docinfo.xml
          - master.adoc
        - qpc
          - dist [symlink]
          - modules [symlink to modules_docs]
          - attributes.adoc
          - docinfo.xml
          - master.adoc
      - docs_gui
        - discovery
          - dist [symlink]
          - modules [symlink to modules_docs]
          - attributes.adoc
          - docinfo.xml
          - master.adoc
        - qpc
          - dist [symlink]
          - modules [symlink to modules_docs]
          - attributes.adoc
          - docinfo.xml
          - master.adoc
      - modules_docs
        - [asciidocs go here]
      - modules_ea
        - [JSON files go here]
    - dist
      - cli
        - index.html
        - index-brand.html
      - gui
        - docs
          - index.html
          - index-brand.html
        - locales
          - en.json
          - locales.json [a list of all available locales in a GUI consumable format]
        - index.js [an interface to allow GUI consumption]
```
- ~We weren't necessarily sure about where to place all the `adoc` files, other than to note @jlprevatt had started using a `-gui` suffix on some changes, those should be located under `src/modules_docs`. We also went ahead and labeled the other aspect `cli` understanding that that label would either change or go away.~
- The `dist` structure is generated. It needs to have the build "manually" run. Next steps are to look at creating some form of basic integration check to call out if that part of the process hasn't been run, or is missing files based on previous checkins, through Jenkins or Travis

### Basic linting check
1. update the NPM packages with `$ npm install` or `$ yarn` (if yarn is installed)
1. `$ npm test` or `$ yarn test`
### Basic build check
Running this should update the `dist` output. Make a change under the `src` directory, run the build script, the output should update.
1. update the NPM packages with `$ npm install` or `$ yarn` (if yarn is installed)
1. `$ npm run build` or `$ yarn build`
1. you can also run each asciidoc build seperate
   - `$ npm run build:discovery-install`
   - `$ npm run build:qpc-install`
   - `$ npm run build:discovery-user`
   - `$ npm run build:qpc-user`
### Basic developing/editing check
What happens when you edit something, or need to work. We went ahead and setup a watch for both the `adoc`s and JSON files to aid in development. ~It runs the build scripts, and linting outputting debugging information as you go.~ It now only runs the asciidoctor scripts because of limitations on the development machines being used.  If you run everything from this PR, right off the line you should see some failures around asciidoctor calling out broken files. The files will still get partially generated to dist, but now an error gets thrown to aid in `CI`... for example
> [build:discovery-install] converting file ./src/docs_install/discovery/master.adoc
> [build:discovery-install] asciidoctor: ERROR: modules/proc-install-docker-from-pkg-centos7.adoc: line 26: invalid part, must have at least one section (e.g., chapter, appendix, etc.)
> [build:discovery-install] asciidoctor: WARNING: modules/proc-install-docker-from-pkg-centos7.adoc: line 27: unterminated listing block
> [build:discovery-install] asciidoctor: ERROR: master.adoc: line 43: invalid part, must have at least one section (e.g., chapter, appendix, etc.)
> [build:discovery-install] error Command failed with exit code 1.

~This is directly related to the fork work we pulled from being a WIP~
1. update the NPM packages with `$ npm install` or `$ yarn` (if yarn is installed)
1. `$ npm start` or `$ yarn start`

## Updates issue/story

- #28 
- pulls in existing PR #23 

@annibond

### CODE REVIEW

**Comment by jlprevatt** [comment]
Met with @cdcabrera  on 05/28/19 and summarized my feedback about the directory structure during that meeting in this doc:
https://docs.google.com/document/d/1_Rxo7hVXswYmi_fg493pTzopEKBv5qigevmljPbJiG0/edit#heading=h.a1n4l52k5gka
David is reviewing the suggestions.

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
+def qpc_version = "0.0.47"
```
Comment (kholdaway) [inline Jenkinsfile]: I don't think this is needed for the build below

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        sh "npm -v"
+        sh "sudo npm install yarn -g"
+        sh "yarn --non-interactive"
+    }
+    stage ('Test Client') {
+        sh "yarn test"
+    }
+    stage('Build Client') {
+        sh "yarn build"
+    }
```
Comment (kholdaway) [inline Jenkinsfile]: Should this produce a build artifact?  Like a dist dir?

Comment (cdcabrera) [reply Jenkinsfile]: Yup, per the off to the side conversation, activating the artifact will become a thing.

The current reason/state it's in it's only running the Asciidoctor setup to confirm everything is good to go build-wise, no errors. A similar setup is in the Travis config.

**Pull request merged**